env_store.go

```diff
@@ -44,41 +52,113 @@ func NewEnvStore() *EnvStore {
 	}
 }
 
-func (s *EnvStore) GetProviderPaths(provider provider.Provider) ([]string, error) {
-	var secretPaths []string
+// GetProviderPaths returns a map of secret paths for each provider
+func (s *EnvStore) GetProviderPaths() map[string][]string {
+	providerPaths := make(map[string][]string)
 
 	for envKey, path := range s.data {
-		p, path := getProviderPath(path)
+		providerName, path := getProviderPath(path)
+		switch providerName {
+		case file.ProviderName:
+			_, ok := providerPaths[file.ProviderName]
+			if !ok {
+				providerPaths[file.ProviderName] = []string{}
+			}
+
+			providerPaths[file.ProviderName] = append(providerPaths[file.ProviderName], path)
+
+		case vault.ProviderName:
+			_, ok := providerPaths[vault.ProviderName]
+			if !ok {
+				providerPaths[vault.ProviderName] = []string{}
+			}
 
-		// TODO(csatib02): Implement multi-provider support
-		if p == provider.GetProviderName() {
 			// The injector function expects a map of key:value pairs
-			if p == vault.ProviderName {
-				path = envKey + "=" + path
+			path = envKey + "=" + path
+			providerPaths[vault.ProviderName] = append(providerPaths[vault.ProviderName], path)
+		}
+	}
+
+	return providerPaths
+}
+
+// GetProviderSecrets creates a new provider for each detected provider using a specified config.
+// It then asynchronously loads secrets using each provider and it's corresponding paths.
+// The secrets from each provider are then placed into a map with the provider name as the key.
+func (s *EnvStore) GetProviderSecrets(providerPaths map[string][]string) (map[string][]provider.Secret, error) {
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	// At most, we will have one error per provider
+	errCh := make(chan error, len(supportedProviders))
+	providerSecrets := make(map[string][]provider.Secret)
+
+	for providerName, paths := range providerPaths {
+		providerSecrets[providerName] = []provider.Secret{}
+		wg.Add(1)
+
```

Comment:
i would not initialize the map for providers at all since we will have to access their values with a safety check anyway. in case one of the providers fails, this result is discarded anyway.

```suggestion
		wg.Add(1)
```